File: backend/app/schemas/outline.py
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.chapter import ChapterStatus


class OutlineSectionOut(BaseModel):
    title: str
    summary: str


class OutlineChapterOut(BaseModel):
    id: UUID
    index: int = Field(..., ge=1)
    title: str
    summary: str | None = None
    key_points: list[str] = []
    estimated_words: int = Field(default=3000, ge=100)
    sections: list[OutlineSectionOut] = []
    word_count: int = 0
    status: ChapterStatus


class OutlineBookResponse(BaseModel):
    title: str
    total_chapters: int
    estimated_words: int
    chapters: list[OutlineChapterOut]


class OutlineChapterPatch(BaseModel):
    index: int = Field(..., ge=1)
    title: str | None = None
    summary: str | None = None
    key_points: list[str] | None = None
    estimated_words: int | None = Field(default=None, ge=100)
    sections: list[OutlineSectionOut] | None = None


class OutlinePut(BaseModel):
    chapters: list[OutlineChapterPatch]
    confirm_start_writing: bool = False


class OutlineGenerateIn(BaseModel):
    topic_override: str | None = Field(default=None, max_length=500)
    target_audience: str | None = Field(default=None, max_length=500)

File: frontend/src/types/outline.ts
import type { ChapterGenStatus } from "./book";

export interface OutlineSection {
  title: string;
  summary: string;
}

export interface OutlineChapter {
  id: string;
  index: number;
  title: string;
  summary: string | null;
  key_points: string[];
  estimated_words: number;
  sections: OutlineSection[];
  word_count: number;
  status: ChapterGenStatus;
}

export interface OutlineBookResponse {
  title: string;
  total_chapters: number;
  estimated_words: number;
  chapters: OutlineChapter[];
}

export interface OutlineChapterPatch {
  index: number;
  title?: string | null;
  summary?: string | null;
  key_points?: string[] | null;
  estimated_words?: number | null;
  sections?: OutlineSection[] | null;
}

export interface OutlinePutPayload {
  chapters: OutlineChapterPatch[];
  confirm_start_writing?: boolean;
}

export interface OutlineGeneratePayload {
  topic_override?: string | null;
  target_audience?: string | null;
}
